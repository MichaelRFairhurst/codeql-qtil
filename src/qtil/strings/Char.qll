private import qtil.inheritance.Instance

private module Str {
  import qtil.strings.Join
}

class Char extends InfInstance<int>::Type {
  bindingset[this]
  Char() { exists(inst().toUnicode()) }

  bindingset[this]
  string toString() { result = inst().toUnicode() }

  bindingset[this]
  int codePoint() { result = inst() }

  bindingset[this]
  predicate isLowercase() { toString().isLowercase() }

  bindingset[this]
  Char toLowercase() { result = toString().toLowerCase().codePointAt(0) }

  bindingset[this]
  predicate isUppercase() { toString().isUppercase() }

  bindingset[this]
  Char toUppercase() { result = toString().toUpperCase().codePointAt(0) }

  bindingset[this]
  predicate isDigit() { toString() in ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"] }

  bindingset[this]
  predicate isAlphabetic() { not toString().toLowerCase() = toString().toUpperCase() }

  bindingset[this]
  predicate isSpecial() {
    not isDigit() and
    not isAlphabetic()
  }

  bindingset[this]
  predicate isAscii() {
    codePoint() >= 0 and
    codePoint() <= 127
  }

  bindingset[s]
  predicate isStr(string s) { s.codePointAt(0) = this }

  bindingset[this, n]
  string repeat(int n) {
    n >= 0 and
    result = concat(int x | x = [1 .. n] | this.toString())
  }

  bindingset[this, s]
  string wrap(string s) { result = this.toString() + s + this.toString() }

  bindingset[this, v1, v2]
  string join(string v1, string v2) { result = Str::join(this.toString(), v1, v2) }

  bindingset[this, v1, v2, v3]
  string join(string v1, string v2, string v3) { result = Str::join(this.toString(), v1, v2, v3) }

  bindingset[this, v1, v2, v3, v4]
  string join(string v1, string v2, string v3, string v4) {
    result = Str::join(this.toString(), v1, v2, v3, v4)
  }

  bindingset[this, v1, v2, v3, v4, v5]
  string join(string v1, string v2, string v3, string v4, string v5) {
    result = Str::join(this.toString(), v1, v2, v3, v4, v5)
  }

  bindingset[this, v1, v2, v3, v4, v5, v6]
  string join(string v1, string v2, string v3, string v4, string v5, string v6) {
    result = Str::join(this.toString(), v1, v2, v3, v4, v5, v6)
  }

  bindingset[this, v1, v2, v3, v4, v5, v6, v7]
  string join(string v1, string v2, string v3, string v4, string v5, string v6, string v7) {
    result = Str::join(this.toString(), v1, v2, v3, v4, v5, v6, v7)
  }

  bindingset[this, v1, v2, v3, v4, v5, v6, v7, v8]
  string join(string v1, string v2, string v3, string v4, string v5, string v6, string v7, string v8) {
    result = Str::join(this.toString(), v1, v2, v3, v4, v5, v6, v7, v8)
  }

  bindingset[this, s, idx]
  string split(string s, int idx) {
    result = s.splitAt(this.toString(), idx)
  }

  bindingset[this, s]
  string split(string s) {
    result = s.splitAt(this.toString())
  }

  bindingset[this, s]
  int indexIn(string s) {
    result = s.indexOf(this.toString())
  }

  bindingset[this, s, startAt]
  int indexIn(string s, int idx, int startAt) {
    result = s.indexOf(this.toString(), idx, startAt)
  }
}

bindingset[s]
Char charOf(string s) { result.isStr(s) }
