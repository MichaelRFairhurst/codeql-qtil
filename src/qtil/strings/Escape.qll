private import qtil.parameterization.SignaturePredicates
private import qtil.tuple.StringTuple
private import qtil.inheritance.Instance
private import qtil.strings.Char

/**
 * An escape map that escapes newlines, carriage returns, and tabs, to `\n`, `\r`, and `\t`
 * respectively.
 *
 * See the `Escape` module for more details on how to use this escape map.
 */
predicate defaultEscapeMap(Char real, Char escaped) {
  real.isStr("\n") and escaped.isStr("n")
  or
  real.isStr("\r") and escaped.isStr("r")
  or
  real.isStr("\t") and escaped.isStr("t")
}

predicate emptyEscapeMap(Char real, Char escaped) { none() }

predicate regexEscapeMap(Char real, Char escaped) {
  // Regex characters do not escape to different characters, but rather to themselves. For example,
  // the newline character escapes to an escaped `n` in most contexts, but a dollar sign is escaped
  // to an escaped dollar sign. An escaped `n` escapes to a newline in most contexts, but an
  // escaped dollar sign in a regex escapes to a plain dollar sign.
  real.isStr(["(", ")", "|", "\\", ".", "+", "*", "?", "^", "$"]) and escaped = real
}

/**
 * A module for building predicates that escape and unescape strings, based on a map of characters
 * to escape and their escaped equivalents in the predicate `escapeMap`.
 *
 * The `escapeMap` predicate is a binary predicate that takes two strings, the first being the
 * character to escape and the second being the escaped equivalent. For example, if a newline should
 * be escaped to `\n`, and `\n` should be unescaped to a newline, then `escapeMap("\n", "n")` should
 * hold.
 *
 * The escape map is designed to work regardless of the escape character used. For example, if the
 * escape character is `$` instead of a backslash, then the previously mentioned escape map would
 * automatically escape `\n` into `$n` and unescape `$n` into `\n`.
 *
 * The predicate `escapeMap` does not need to declare that the escape character should itself be
 * escaped. This is handled by the `escape` and `unescape` predicates.
 *
 * ```ql
 * predicate myEscapeMap(Char real, Char escaped) {
 *  real.isStr("\n") and escaped.isStr("n")
 * }
 *
 * // Selects "foo\\nbar\\\\baz", "foo\nbar\\baz".
 * select Escape<myEscapeMap>::escape("foo\nbar\\baz", "\\"),
 *   Escape<myEscapeMap>::unescape("foo\\nbar\\\\baz", "\\")
 *
 * // Selects "foo$nbar$$baz", "foo\nbar$baz".
 * select Escape<myEscapeMap>::escape("foo\nbar$baz", "$"),
 *   Escape<myEscapeMap>::unescape("foo$nbar$$baz", "$")
 * ```
 *
 * The following predefined escape maps are available:
 * - `defaultEscapeMap`: The default escape map, which escapes newlines, carriage returns, and tabs.
 * - `regexEscapeMap`: The regex escape map, which escapes all regex special characters.
 * - `emptyEscapeMap`: An empty escape map, which does not escape any characters.
 */
module Escape<Binary<Char, Char>::pred/2 escapeMap> {
  /**
   * Perform the escape operation on a string, using the escape map, and a custom escape character.
   *
   * The escape character is also escaped in the string.
   *
   * See the module documentation (`Escape`) for more details.
   */
  bindingset[input, escaper]
  string escape(string input, Char escaper) {
    result =
      concat(int cpIn, int cpIdx, string strOut |
        cpIn = input.codePointAt(cpIdx) and
        if replacesOnEscape(input, escaper, cpIdx, _)
        then replacesOnEscape(input, escaper, cpIdx, strOut)
        else strOut = cpIn.toUnicode()
      |
        strOut order by cpIdx
      )
  }

  /**
   * Perform the escape operation on a string, using the escape map, using backslash as the escape
   * character.
   *
   * Backslash characters are also escaped in the string.
   *
   * See the module documentation (`Escape`) for more details.
   */
  bindingset[input]
  string escape(string input) { result = escape(input, charOf("\\")) }

  /**
   * Reverse the escape operation on a string, using the escape map, and a custom escape character.
   *
   * The escape character is also unescaped in the string when it is used to escape itself.
   *
   * See the module documentation (`Escape`) for more details.
   */
  bindingset[input, escaper]
  string unescape(string input, Char escaper) {
    result =
      concat(int groupIdx, string groupText, string strOut |
        groupText =
          input
              .regexpFind("(\\" + escaper.toString() + ".|[^\\" + escaper.toString() + "])",
                groupIdx, _) and
        unescapeReplaces(escaper, groupText, strOut)
      |
        strOut order by groupIdx
      )
  }

  /**
   * Reverse the escape operation on a string, using the escape map, using backslash as the escape
   * character.
   *
   * Double backslash characters are also unescaped in the string to a single backslash.
   *
   * See the module documentation (`Escape`) for more details.
   */
  bindingset[input]
  string unescape(string input) { result = unescape(input, charOf("\\")) }

  /**
   * Given the escape map, holds for an escaper, and a character or two character group, and the
   * replacement for that group.
   *
   * For example, if the escape map is `escapeMap("\n", "n")`, and the escaper is `\`, then
   * `unescapeReplaces("\\", "\\n", "\n")` holds, and `unescapeReplaces("\\", "a", "a") holds.
   *
   * This is because the unescape algorithm is just a state machine (implemented as a regex) that
   * grabs one character at a time, unless it sees the escaper, in which case it grabs two. Single
   * characters are not transformed. Escaped characters are transformed according to the escape
   * map: many characters are unescaped to themselves, but some are unescaped to other characters.
   * For example, a double backslash is typically unescaped to a single backslash, and an escaped
   * `n` is typically unescaped to a newline character.
   */
  bindingset[escaper, group]
  private predicate unescapeReplaces(Char escaper, string group, string output) {
    group = escaper.repeat(2) and
    output = escaper.toString()
    or
    exists(Char out, Char inp |
      escapeMap(out, inp) and
      group = escaper.toString() + inp.toString() and
      output = out.toString()
    )
    or
    not group.codePointAt(0) = escaper and
    output = group
  }

  /**
   * Given the escape map, holds for an escaper, a character index in the input string, and the
   * replacement for that character.
   *
   * For example, if the escape map is `escapeMap("\n", "n")`, and the escaper is `\`, then
   * `replacesOnEscape("foo\nbar", "\\", 3, "\\n")` holds, and
   * `replacesOnEscape("foo\\nbar", "\\", 0, "f")` holds.
   *
   * This performs a simple job. If we see an escape character, we escape it. If we see an escape
   * mapped character (a -> b), we place an escape before the new character "b". Anything else is
   * left alone.
   *
   * Note that many escape maps escape to themselves. For example, a double quote is typically
   * escaped by adding the escape character before it, and otherwise the double quote is not
   * changed. However, a newline character is typically not escaped to a backslash before a newline
   * character, but rather to a backslash followed by an `n` character.
   */
  bindingset[input, escaper, cpIdx]
  private predicate replacesOnEscape(string input, Char escaper, int cpIdx, string output) {
    exists(Char codeOut |
      escapeMap(input.codePointAt(cpIdx), codeOut) and
      output = escaper.toString() + codeOut.toString()
    )
    or
    input.codePointAt(cpIdx) = escaper and
    output = escaper.repeat(2)
  }
}

/**
 * A module for wrapping a string with a character, and escaping the wrapping character in the
 * string.
 *
 * For instance, this can wrap a string with double quotes, and escape the double quotes already in
 * the string.
 *
 * Takes an escape map predicate `escapeMap` that defines additional escaping behavior, such as
 * turning newlines into an escaped `n`.
 */
module WrapEscape<Nullary::Ret<Char>::pred/0 wrapChar, Binary<Char, Char>::pred/2 escapeMap> {
  /**
   * Add the wrapping character to the escape map, so that it is escaped to itself.
   */
  predicate newEscapeMap(Char real, Char escaped) {
    real = wrapChar() and escaped = wrapChar()
    or
    escapeMap(real, escaped)
  }

  /**
   * Escape a string using the escape map, and wrap the result with the wrapping character.
   *
   * The wrapping character, and the escape character, are also both escaped in the string.
   *
   * For example:
   *
   * ```ql
   * import WrapEscaping<Separator::percent/0, defaultEscapeMap>
   *
   * // Selects "%foo\\%bar\\\\baz" and "%foo\\%bar\\\\baz\\nqux".
   * select wrapEscaping("foo%bar\\baz", "\\"), wrapEscaping("foo%bar\\baz\nqux", "\\")
   *
   * // Selects "%foo.%bar..baz" and "%foo%bar..baz.nqux".
   * select wrapEscaping("foo%bar$baz", "."), wrapEscaping("foo%bar$baz\nqux", ".")
   * ```
   */
  bindingset[str, escape]
  string wrapEscaping(string str, Char escape) {
    result = wrapChar().wrap(Escape<newEscapeMap/2>::escape(str, escape))
  }

  bindingset[str]
  string wrapEscaping(string str) { result = wrapEscaping(str, charOf("\\")) }

  bindingset[str, escape]
  string unwrapUnescaping(string str, Char escape) {
    exists(string unwrapped |
      // Cut out the wrapping characters.
      unwrapped = str.substring(1, str.length() - 1) and
      // Ensure the character was actually wrapped with the wrapping character.
      str = wrapChar().wrap(unwrapped) and
      // Unescape the contents between the wrapping characters.
      result = Escape<newEscapeMap/2>::unescape(unwrapped, escape)
    )
  }

  bindingset[str]
  string unwrapUnescaping(string str) { result = unwrapUnescaping(str, charOf("\\")) }
}

/** Predicate to parameterize `WrapEscape` for double quoting strings. */
private Char doubleQuote() { result.isStr("\"") }

/** Predicate to parameterize `WrapEscape` for single quoting strings. */
private Char singleQuote() { result.isStr("'") }

/**
 * Wrap a string with double quotes, adding a backslash escape character if necessary.
 *
 * For instance, `doubleQuoteEscaping("foo") = "\"foo\""`, and
 * `doubleQuoteEscaping("foo\"bar") = "\"foo\\\"bar\""`.
 */
bindingset[str]
string doubleQuoteEscaping(string str) {
  result = WrapEscape<doubleQuote/0, defaultEscapeMap/2>::wrapEscaping(str, charOf("\\"))
}

/**
 * Unwrap a string with double quotes, removing the double quotes and any escape characters.
 *
 * For instance, `unescapeDoubleQuote("\"foo\"") = "foo"`, and
 * `unescapeDoubleQuote("\"foo\\\"bar\"") = "foo\"bar"`.
 *
 * Does not hold for any result value in the case where the string is not double quoted.
 */
bindingset[str]
string unescapeDoubleQuote(string str) {
  result = WrapEscape<doubleQuote/0, defaultEscapeMap/2>::unwrapUnescaping(str, charOf("\\"))
}

/**
 * Wrap a string with single quotes, adding a backslash escape character if necessary.
 *
 * For instance, `singleQuoteEscaping("foo") = "'foo'"`, and
 * `singleQuoteEscaping("foo'bar") = "'foo\\'bar"`.
 */
bindingset[str]
string singleQuoteEscaping(string str) {
  result = WrapEscape<singleQuote/0, defaultEscapeMap/2>::wrapEscaping(str, charOf("\\"))
}

/**
 * Unwrap a string with single quotes, removing the single quotes and any escape characters.
 *
 * For instance, `unescapeSingleQuote("'foo'") = "foo"`, and
 * `unescapeSingleQuote("'foo\\'bar'") = "foo'bar"`.
 *
 * Does not hold for any result value in the case where the string is not single quoted.
 */
bindingset[str]
string unescapeSingleQuote(string str) {
  result = WrapEscape<singleQuote/0, defaultEscapeMap/2>::unwrapUnescaping(str, charOf("\\"))
}

/**
 * Add escapes to a string literal to make it safe for use in a Java-style PRCE regex, as a direct
 * character by character match.
 *
 * For instance, `regexEscape("foo") = "foo"`, and
 * `regexEscape("^foo(bar)baz[qux]$") = "\\^foo\\(bar\\)baz\\[qux\\]\\$"`.
 */
bindingset[str]
string escapeRegex(string str) { result = Escape<regexEscapeMap/2>::escape(str, charOf("\\")) }

bindingset[str]
string unescapeRegex(string str) { result = Escape<regexEscapeMap/2>::unescape(str, charOf("\\")) }
