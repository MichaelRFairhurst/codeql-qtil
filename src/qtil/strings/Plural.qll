bindingset[str]
string plural(string str) { result = str + "s" }

bindingset[str, pcount]
string plural(string str, int pcount) { result = plural(str, plural(str), pcount) }

bindingset[single, plural, pcount]
string plural(string single, string plural, int pcount) {
  if pcount = 1 then result = single else result = plural
}
