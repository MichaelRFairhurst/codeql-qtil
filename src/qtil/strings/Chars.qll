import qtil.strings.Char

/**
 * A module that declares predicates for defining common separators to the various `ListBuilder`
 * modules defined in this file.
 */
module Chars {
  Char comma() { result.isStr(",") }

  Char dollar() { result.isStr("$") }

  Char colon() { result.isStr(":") }

  Char at() { result.isStr("@") }

  Char hash() { result.isStr("#") }

  Char excl() { result.isStr("!") }

  Char caret() { result.isStr("^") }

  Char amp() { result.isStr("&") }

  Char pipe() { result.isStr("|") }

  Char semicolon() { result.isStr(";") }

  Char plus() { result.isStr("+") }

  Char minus() { result.isStr("-") }

  Char slash() { result.isStr("/") }

  Char backslash() { result.isStr("\\") }

  Char dot() { result.isStr(".") }

  Char question() { result.isStr("?") }

  Char percent() { result.isStr("%") }

  Char tilde() { result.isStr("~") }

  Char space() { result.isStr(" ") }

  Char tab() { result.isStr("\t") }

  Char newline() { result.isStr("\n") }

  Char backtick() { result.isStr("`") }
}