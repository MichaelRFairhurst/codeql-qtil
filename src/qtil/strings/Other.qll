import qtil.parameterization.SignaturePredicates
import qtil.parameterization.SignatureTypes
import qtil.strings.Char

/**
 * CodeQL 2.21.1 has a bug where `concat` with a delimiter and an order will silently drop the
 * delimiter.
 *
 * Use this module to work around that bug.
 *
 * Example:
 *
 * ```
 * string names(int x) {
 *   x = 0 and result = "foo" or
 *   x = 1 and result = "bar" or
 *   x = 2 and result = "baz"
 * }
 *
 * // BUG: selects foobarbaz
 * select concat(int x, string name | name = names(x) | name order by x, ",")
 *
 * // FIXED: selects foo,bar,baz
 * select ConcatDelimOrderFixed<Chars::comma/0, names/1>::join()
 * ```
 */
module ConcatDelimOrderFixed<
  Nullary::Ret<Char>::pred/0 separator, Unary<int>::Ret<string>::pred/1 items>
{
  /**
   * Perform the concatenation with the given separator and items/order.
   */
  string join() {
    exists(int lastIdx |
      lastIdx = max(int x | exists(items(x))) and
      result =
        concat(int x, string given, string separated |
          given = items(x) and
          if x = lastIdx then separated = given else separated = given + separator()
        |
          separated order by x
        )
    )
    or
    // Match CodeQL concat() behavior and always return an empty string if there are no items.
    not exists(items(_)) and result = ""
  }
}
