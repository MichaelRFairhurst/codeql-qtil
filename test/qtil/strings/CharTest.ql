/**
 * WARNING: This test is incredibly slow to compile.
 * 
 * CodeQL gets into some pathological cases and it may take over a minute to compile, even though
 * it runs in about a hundred milliseconds.
 */

import qtil.strings.Char
import qtil.testing.Qnit

class TestCharToStringIsStr extends Test, Case {
  override predicate run(Qnit test) {
    if
      charOf("a").isStr("a") and
      charOf("b").isStr("b") and
      charOf("c").isStr("c") and
      charOf("d").isStr("d") and
      charOf("e").isStr("e") and
      charOf("f").isStr("f") and
      charOf("g").isStr("g") and
      charOf("h").isStr("h") and
      charOf("i").isStr("i") and
      charOf("j").isStr("j") and
      charOf("$").isStr("$") and
      charOf("!").isStr("!") and
      charOf(" ").isStr(" ") and
      charOf("~").isStr("~") and
      charOf("1").isStr("1") and
      charOf("2").isStr("2") and
      charOf("3").isStr("3") and
      charOf("\n").isStr("\n") and
      charOf("\t").isStr("\t")
    then test.pass("Char to string isStr works")
    else test.fail("Char to string isStr doesn't work")
  }
}

class TestCharToString extends Test, Case {
  override predicate run(Qnit test) {
    if
      charOf("a").toString() = "a" and
      charOf("b").toString() = "b" and
      charOf("c").toString() = "c" and
      charOf("d").toString() = "d" and
      charOf("e").toString() = "e" and
      charOf("f").toString() = "f" and
      charOf("g").toString() = "g" and
      charOf("h").toString() = "h" and
      charOf("i").toString() = "i" and
      charOf("j").toString() = "j" and
      charOf("$").toString() = "$" and
      charOf("!").toString() = "!" and
      charOf(" ").toString() = " " and
      charOf("~").toString() = "~" and
      charOf("1").toString() = "1" and
      charOf("2").toString() = "2" and
      charOf("3").toString() = "3" and
      charOf("\n").toString() = "\n" and
      charOf("\t").toString() = "\t"
    then test.pass("Char to string works")
    else test.fail("Char to string doesn't work")
  }
}

class TestCharToStringIsNotStr extends Test, Case {
  override predicate run(Qnit test) {
    if
      charOf("a")
          .isStr([
              "A", "b", "c", "d", "e", "f", "g", "h", "i", "j", "$", "!", " ", "~", "1", "2", "3",
              "\n", "\t"
            ]) or
      charOf(" ")
          .isStr([
              "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "$", "!", "~", "1", "2", "3", "\n",
              "\t"
            ]) or
      charOf("1")
          .isStr(["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "$", "!", "~", " ", "\n", "\t"]) or
      charOf("$")
          .isStr(["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "!", "~", " ", "\n", "\t"])
    then test.fail("Char to string isStr holding for incorrect values")
    else test.pass("Char to string isStr correctly not holding for incorrect values")
  }
}

class TestCharIsDigit extends Test, Case {
  override predicate run(Qnit test) {
    if
      charOf(["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]).isDigit() and
      not charOf(["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "$", "!", " ", "~", "\n", "\t"])
          .isDigit()
    then test.pass("Char is digit works")
    else test.fail("Char is digit doesn't work")
  }
}

class TestCharIsAlphabetic extends Test, Case {
  override predicate run(Qnit test) {
    if
      charOf([
          "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r",
          "s", "t", "u", "v", "w", "x", "y", "z"
        ]).isAlphabetic() and
      charOf([
          "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R",
          "S", "T", "U", "V", "W", "X", "Y", "Z"
        ]).isAlphabetic() and
      not charOf(["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "$", "!", " ", "~", "\n", "\t"])
          .isAlphabetic()
    then test.pass("Char is alphabetic works")
    else test.fail("Char is alphabetic doesn't work")
  }
}

class TestCharIsSpecial extends Test, Case {
  override predicate run(Qnit test) {
    if
      charOf(["$", "!", " ", "~", "\n", "\t"]).isSpecial() and
      not charOf([
          "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r",
          "s", "t", "u", "v", "w", "x", "y", "z"
        ]).isSpecial() and
      not charOf(["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]).isSpecial()
    then test.pass("Char is special works")
    else test.fail("Char is special doesn't work")
  }
}

class TestCharIsLowercase extends Test, Case {
  override predicate run(Qnit test) {
    if
      charOf([
          "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r",
          "s", "t", "u", "v", "w", "x", "y", "z"
        ]).isLowercase() and
      not charOf([
          "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R",
          "S", "T", "U", "V", "W", "X", "Y", "Z"
        ]).isLowercase() and
      // Match string.isLowercase() behavior: if its not an uppercase character, it is considered lowercase.
      charOf(["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]).isLowercase() and
      charOf(["$", "!", " ", "~", "\n", "\t"]).isLowercase()
    then test.pass("Char is lowercase works")
    else test.fail("Char is lowercase doesn't work")
  }
}

class TestCharIsUppercase extends Test, Case {
  override predicate run(Qnit test) {
    if
      charOf([
          "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R",
          "S", "T", "U", "V", "W", "X", "Y", "Z"
        ]).isUppercase() and
      not charOf(["a", "b", "c", "d", "e", "f", "g", "h", "i", "j"]).isUppercase() and
      // Match string.isUppercase() behavior: if its not a lowercase character, it is considered uppercase.
      charOf(["0", "1", "2", "3", "4", "5", "6", "7"]).isUppercase() and
      charOf(["$", "$"]).isUppercase()
    then test.pass("Char is uppercase works")
    else test.fail("Char is uppercase doesn't work")
  }
}

class TestCharIsAscii extends Test, Case {
  override predicate run(Qnit test) {
    if
      [0 .. 127].(Char).isAscii() and
      not [128 .. 526].(Char).isAscii()
    then test.pass("Char is ascii works")
    else test.fail("Char is ascii doesn't work")
  }
}

class TestCharCodePoint extends Test, Case {
  override predicate run(Qnit test) {
    if
      charOf("a").codePoint() = 97 and
      charOf("b").codePoint() = 98 and
      charOf("c").codePoint() = 99 and
      charOf("d").codePoint() = 100 and
      charOf("e").codePoint() = 101 and
      charOf("f").codePoint() = 102 and
      charOf("g").codePoint() = 103 and
      charOf("h").codePoint() = 104 and
      charOf("i").codePoint() = 105 and
      charOf("j").codePoint() = 106 and
      charOf("0").codePoint() = 48 and
      charOf("1").codePoint() = 49 and
      charOf("2").codePoint() = 50 and
      charOf("3").codePoint() = 51 and
      charOf("$").codePoint() = 36 and
      charOf("!").codePoint() = 33 and
      charOf(" ").codePoint() = 32 and
      charOf("~").codePoint() = 126 and
      charOf("\n").codePoint() = 10 and
      charOf("\t").codePoint() = 9
    then test.pass("Char code point works")
    else test.fail("Char code point doesn't work" + charOf("a").codePoint().toString())
  }
}

class TestIntToChar extends Test, Case {
  override predicate run(Qnit test) {
    if
      97.(Char).isStr("a") and
      98.(Char).isStr("b") and
      99.(Char).isStr("c") and
      100.(Char).isStr("d") and
      101.(Char).isStr("e") and
      102.(Char).isStr("f") and
      103.(Char).isStr("g") and
      104.(Char).isStr("h") and
      105.(Char).isStr("i") and
      106.(Char).isStr("j") and
      48.(Char).isStr("0") and
      49.(Char).isStr("1") and
      50.(Char).isStr("2") and
      51.(Char).isStr("3") and
      36.(Char).isStr("$") and
      33.(Char).isStr("!") and
      32.(Char).isStr(" ") and
      126.(Char).isStr("~") and
      10.(Char).isStr("\n") and
      9.(Char).isStr("\t")
    then test.pass("Int to char works")
    else test.fail("Int to char doesn't work")
  }
}

class TestToUpperCase extends Test, Case {
  override predicate run(Qnit test) {
    if
      charOf(["a", "A"]).toUppercase().isStr("A") and
      charOf(["b", "B"]).toUppercase().isStr("B") and
      charOf(["c", "C"]).toUppercase().isStr("C") and
      // Optimizer seems to hit a pathological case here and doesn't terminate.
      /*
       * charOf(["d", "D"]).toUppercase().isStr("D") and
       *      charOf(["e", "E"]).toUppercase().isStr("E") and
       *      charOf(["f", "F"]).toUppercase().isStr("F") and
       *      charOf(["g", "G"]).toUppercase().isStr("G") and
       *      charOf(["h", "H"]).toUppercase().isStr("H") and
       *      charOf(["i", "I"]).toUppercase().isStr("I") and
       *      charOf(["j", "J"]).toUppercase().isStr("J") and
       *      charOf(["k", "K"]).toUppercase().isStr("K") and
       *      charOf(["l", "L"]).toUppercase().isStr("L") and
       *      charOf(["m", "M"]).toUppercase().isStr("M") and
       *      charOf(["n", "N"]).toUppercase().isStr("N") and
       *      charOf(["o", "O"]).toUppercase().isStr("O") and
       *      charOf(["p", "P"]).toUppercase().isStr("P") and
       *      charOf(["q", "Q"]).toUppercase().isStr("Q") and
       *      charOf(["r", "R"]).toUppercase().isStr("R") and
       *      charOf(["s", "S"]).toUppercase().isStr("S") and
       *      charOf(["t", "T"]).toUppercase().isStr("T") and
       *      charOf(["u", "U"]).toUppercase().isStr("U") and
       *      charOf(["v", "V"]).toUppercase().isStr("V") and
       *      charOf(["w", "W"]).toUppercase().isStr("W") and
       *      charOf(["x", "X"]).toUppercase().isStr("X") and
       */

      charOf(["y", "Y"]).toUppercase().isStr("Y") and
      charOf(["z", "Z"]).toUppercase().isStr("Z") and
      charOf("0").toUppercase().isStr("0") and
      charOf("1").toUppercase().isStr("1") and
      charOf("2").toUppercase().isStr("2") and
      charOf("$").toUppercase().isStr("$") and
      charOf("!").toUppercase().isStr("!") and
      charOf(" ").toUppercase().isStr(" ") and
      charOf("\n").toUppercase().isStr("\n")
    then test.pass("Char to uppercase works")
    else test.fail("Char to uppercase doesn't work")
  }
}

class TestToLowerCase extends Test, Case {
  override predicate run(Qnit test) {
    if
      charOf(["a", "A"]).toLowercase().isStr("a") and
      charOf(["b", "B"]).toLowercase().isStr("b") and
      charOf(["c", "C"]).toLowercase().isStr("c") and
      /*
       * Optimizer seems to hit a pathological case here and doesn't terminate.
       *      charOf(["d", "D"]).toLowercase().isStr("d") and
       *      charOf(["e", "E"]).toLowercase().isStr("e") and
       *      charOf(["f", "F"]).toLowercase().isStr("f") and
       *      charOf(["g", "G"]).toLowercase().isStr("g") and
       *      charOf(["h", "H"]).toLowercase().isStr("h") and
       *      charOf(["i", "I"]).toLowercase().isStr("i") and
       *      charOf(["j", "J"]).toLowercase().isStr("j") and
       *      charOf(["k", "K"]).toLowercase().isStr("k") and
       *      charOf(["l", "L"]).toLowercase().isStr("l") and
       *      charOf(["m", "M"]).toLowercase().isStr("m") and
       *      charOf(["n", "N"]).toLowercase().isStr("n") and
       *      charOf(["o", "O"]).toLowercase().isStr("o") and
       *      charOf(["p", "P"]).toLowercase().isStr("p") and
       *      charOf(["q", "Q"]).toLowercase().isStr("q") and
       *      charOf(["r", "R"]).toLowercase().isStr("r") and
       *      charOf(["s", "S"]).toLowercase().isStr("s") and
       *      charOf(["t", "T"]).toLowercase().isStr("t") and
       *      charOf(["u", "U"]).toLowercase().isStr("u") and
       *      charOf(["v", "V"]).toLowercase().isStr("v") and
       *      charOf(["w", "W"]).toLowercase().isStr("w") and
       *      charOf(["x", "X"]).toLowercase().isStr("x") and
       */

      charOf(["y", "Y"]).toLowercase().isStr("y") and
      charOf(["z", "Z"]).toLowercase().isStr("z") and
      charOf("0").toLowercase().isStr("0") and
      charOf("1").toLowercase().isStr("1") and
      charOf("2").toLowercase().isStr("2") and
      charOf("$").toLowercase().isStr("$") and
      charOf("!").toLowercase().isStr("!") and
      charOf(" ").toLowercase().isStr(" ") and
      charOf("\n").toLowercase().isStr("\n")
    then test.pass("Char to lowercase works")
    else test.fail("Char to lowercase doesn't work")
  }
}

class TestCharRepeat extends Test, Case {
  override predicate run(Qnit test) {
    if
      not exists(charOf("a").repeat(-1)) and
      charOf("a").repeat(0) = "" and
      charOf("a").repeat(1) = "a" and
      charOf("b").repeat(2) = "bb" and
      charOf("c").repeat(3) = "ccc" and
      charOf("d").repeat(4) = "dddd" and
      charOf("e").repeat(5) = "eeeee" and
      charOf("f").repeat(6) = "ffffff" and
      charOf("g").repeat(7) = "ggggggg" and
      charOf("h").repeat(8) = "hhhhhhhh" and
      charOf("i").repeat(9) = "iiiiiiiii" and
      charOf("j").repeat(10) = "jjjjjjjjjj" and
      charOf("0").repeat(1) = "0" and
      charOf("1").repeat(2) = "11" and
      charOf("2").repeat(3) = "222" and
      charOf("3").repeat(4) = "3333" and
      charOf("$").repeat(5) = "$$$$$" and
      charOf("!").repeat(6) = "!!!!!!" and
      charOf(" ").repeat(7) = "       " and
      charOf("~").repeat(8) = "~~~~~~~~" and
      charOf("\n").repeat(9) = "\n\n\n\n\n\n\n\n\n" and
      charOf("\t").repeat(10) = "\t\t\t\t\t\t\t\t\t\t"
    then test.pass("Char repeat works")
    else test.fail("Char repeat doesn't work")
  }
}

class TestCharWrap extends Test, Case {
  override predicate run(Qnit test) {
    if
      charOf("a").wrap("foo") = "afooa" and
      charOf("'").wrap("bar") = "'bar'" and
      charOf("\"").wrap("baz") = "\"baz\""
    then test.pass("Char wrap works")
    else test.fail("Char wrap doesn't work")
  }
}
