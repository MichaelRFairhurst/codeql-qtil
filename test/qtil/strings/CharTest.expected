| All 15 tests passed. |
